master/internal/provisioner/provisioner.go

```diff
@@ -110,18 +111,14 @@ func (p *Provisioner) provision(ctx *actor.Context) {
 	}
 
 	ctx.Log().Debug("scale happening")
-	toTerminate := p.scaleDecider.findInstancesToTerminate(
-		ctx, p.provider.maxInstanceNum(),
-	)
-	if len(toTerminate) > 0 {
-		p.provider.terminate(ctx, toTerminate)
+	if toTerminate := p.scaleDecider.findInstancesToTerminate(); len(toTerminate.InstanceIDs) > 0 {
+		ctx.Log().Infof("terminating %d instances: %s",
+			len(toTerminate.InstanceIDs), toTerminate.String())
+		p.provider.terminate(ctx, toTerminate.InstanceIDs)
 	}
 
-	numToLaunch := p.scaleDecider.calculateNumInstancesToLaunch(
-		p.provider.instanceType(),
-		p.provider.maxInstanceNum(),
-	)
-	if numToLaunch > 0 {
+	if numToLaunch := p.scaleDecider.calculateNumInstancesToLaunch(); numToLaunch > 0 {
+		ctx.Log().Infof("launching %d instances", numToLaunch)
```

Comment:
Similarly here, this log line might be misleading. In particular in the case of spot instances, we aren't actually "launching instances" during `launch()`, we are submitting a spot request, which may or may not launch an instance.

Reply:
same

Reply:
Sounds good to me